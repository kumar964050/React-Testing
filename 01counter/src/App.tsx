import { useState } from "react";

import Button from "./components/ui/Button";
import "./App.css";

function App() {
  const [count, setCount] = useState(0);

  const handleIncrease = () => setCount(count + 1);
  const handleDecrease = () => setCount((prev) => (prev > 0 ? prev - 1 : prev));
  const handleReset = () => setCount(0);

  return (
    <>
      <div className="card">
        <h1>Counter App</h1>
        <h2>
          count is <span data-testid="counter-value">{count}</span>
        </h2>
        <Button onClick={handleDecrease}>Decrease</Button>
        <Button onClick={handleReset}>Reset</Button>
        <Button onClick={handleIncrease}>Increase</Button>
      </div>
    </>
  );
}

export default App;
