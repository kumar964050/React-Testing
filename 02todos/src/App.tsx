import React, { useState } from "react";
import { v4 as uuidV4 } from "uuid";
import type { Todo } from "./types";

import TodoItem from "./components/TodoItem";
import Button from "./components/Button";
import Input from "./components/Input";

const App = () => {
  const [text, setText] = useState<string>("");
  const [todos, setTodos] = useState<Todo[]>([]);

  // add new item to the store
  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    // new todo
    const newTodo: Todo = {
      id: uuidV4(),
      text: text.trim(),
      isCompleted: false,
    };
    setTodos((prev: Todo[]) => [...prev, newTodo]);
    setText("");
  };
  const handleSave = () => {
    alert("save btn triggered");
  };
  const toggleTodoStatus = (id: string, value?: boolean): void => {
    setTodos((prev: Todo[]) =>
      prev.map((todo: Todo) => {
        if (todo.id === id)
          return { ...todo, isCompleted: value ?? !todo.isCompleted };
        return todo;
      })
    );
  };
  const handleDelete = (id: string): void => {
    setTodos((prev: Todo[]) => prev.filter((todo: Todo) => todo.id !== id));
  };

  return (
    <div className="todos-bg-container">
      <div className="container">
        <div className="row">
          <div className="col-12">
            <h1 className="todos-heading">Todos </h1>
            <h1 className="create-task-heading create-task-heading-subpart">
              Create Task
            </h1>
            <form onSubmit={handleSubmit}>
              <Input
                type="text"
                className="todo-user-input"
                placeholder="What needs to be done?"
                value={text}
                onChange={(e) => setText(e.target.value)}
              />
              <Button>Add</Button>
            </form>
            <h1 className="todo-items-heading">
              My <span className="todo-items-heading-subpart">Tasks</span>
            </h1>
            <ul className="todo-items-container">
              {todos.map((todo) => (
                <TodoItem
                  key={todo.id}
                  {...todo}
                  handleDelete={handleDelete}
                  toggleTodoStatus={toggleTodoStatus}
                />
              ))}
            </ul>
            <Button onClick={handleSave}>Save</Button>
          </div>
        </div>
      </div>
    </div>
  );
};

export default App;
