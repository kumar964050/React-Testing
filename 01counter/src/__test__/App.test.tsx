import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import App from "../App";

describe("App Component", () => {
  it("renders heading", () => {
    render(<App />);
    expect(
      screen.getByRole("heading", { name: /counter app/i })
    ).toBeInTheDocument();
  });

  it("should start with counter value 0", () => {
    render(<App />);
    expect(screen.getByTestId("counter-value")).toHaveTextContent("0");
  });

  it("increments counter on + button click", async () => {
    render(<App />);
    const button = screen.getByRole("button", { name: /increase/i });
    const value = screen.getByTestId("counter-value");

    await userEvent.click(button);
    expect(value).toHaveTextContent("1");
  });

  it("decrements counter on - button click", async () => {
    render(<App />);
    const button = screen.getByRole("button", { name: /decrease/i });
    const value = screen.getByTestId("counter-value");

    await userEvent.click(button);
    // assumes restriction at 0
    expect(value).toHaveTextContent("0");
  });

  it("resets counter to 0", async () => {
    render(<App />);
    const incBtn = screen.getByRole("button", { name: /increase/i });
    const resetBtn = screen.getByRole("button", { name: /reset/i });
    const value = screen.getByTestId("counter-value");

    await userEvent.click(incBtn);
    await userEvent.click(incBtn);
    expect(value).toHaveTextContent("2");

    await userEvent.click(resetBtn);
    expect(value).toHaveTextContent("0");
  });

  it("handles rapid clicks correctly", async () => {
    render(<App />);
    const button = screen.getByRole("button", { name: /increase/i });
    const value = screen.getByTestId("counter-value");

    await userEvent.dblClick(button); // double click
    expect(value).toHaveTextContent("2");

    await userEvent.click(button);
    expect(value).toHaveTextContent("3");
  });

  it("should not go below 0 if restricted", async () => {
    render(<App />);
    const decButton = screen.getByRole("button", { name: /decrease/i });
    const value = screen.getByTestId("counter-value");

    await userEvent.click(decButton);
    expect(Number(value.textContent)).toBeGreaterThanOrEqual(0);
  });

  it("allows keyboard interaction", async () => {
    render(<App />);
    const button = screen.getByRole("button", { name: /increase/i });
    const value = screen.getByTestId("counter-value");

    button.focus();
    await userEvent.keyboard("{Enter}");
    expect(value).toHaveTextContent("1");
  });

  it("matches snapshot", () => {
    const { asFragment } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
  });
});
