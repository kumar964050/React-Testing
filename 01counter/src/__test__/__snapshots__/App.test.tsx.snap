// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App Component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <h1>
      Counter App
    </h1>
    <h2>
      count is 
      <span
        data-testid="counter-value"
      >
        0
      </span>
    </h2>
    <button
      class="px-4 py-2 rounded bg-blue-500 text-white hover:bg-blue-600 disabled:bg-gray-400"
    >
      Decrease
    </button>
    <button
      class="px-4 py-2 rounded bg-blue-500 text-white hover:bg-blue-600 disabled:bg-gray-400"
    >
      Reset
    </button>
    <button
      class="px-4 py-2 rounded bg-blue-500 text-white hover:bg-blue-600 disabled:bg-gray-400"
    >
      Increase
    </button>
  </div>
</DocumentFragment>
`;
